### PR TITLE
Feature/Branches-meals

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: app/views/branches/_branch_meal.html.erb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        <% end %>
+      <% else %>
+        <%= link_to toggle_meal_active_branch_path(branch, branch_meal_id: branch_meal.id), class: 'btn btn-danger btn-sm me-2', role: 'button' do %>
+          <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-eye-slash-fill" viewBox="0 0 16 16">
+            <path d="m10.79 12.912-1.614-1.615a3.5 3.5 0 0 1-4.474-4.474l-2.06-2.06C.938 6.278 0 8 0 8s3 5.5 8 5.5a7.029 7.029 0 0 0 2.79-.588zM5.21 3.088A7.028 7.028 0 0 1 8 2.5c5 0 8 5.5 8 5.5s-.939 1.721-2.641 3.238l-2.062-2.062a3.5 3.5 0 0 0-4.474-4.474L5.21 3.089z"/>
+            <path d="M5.525 7.646a2.5 2.5 0 0 0 2.829 2.829l-2.83-2.829zm4.95.708-2.829-2.83a2.5 2.5 0 0 1 2.829 2.829zm3.171 6-12-12 .708-.708 12 12-.708.708z"/>
+          </svg>
+        <% end %>
+      <% end %>
+      <%= link_to 'Edit', edit_meal_path(branch_meal.meal), class: 'btn btn-primary btn-sm me-2' %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/branches/_branch_meal.html.erb:26]: remove edit and delete button from here
Add a message saying this page is just to enable disable meals in a branch
to make changes visit meals page(make this link)

File: app/controllers/branches_controller.rb [hunk 2]
```diff
@@ -34,11 +35,54 @@ def destroy
...
+    redirect_to meals_branch_path(params[:id])
+  end
+
+  def toggle_meal_inactive
+    @branch_meal.update(active: false)
+    redirect_to meals_branch_path(params[:id])
+  end
+
+  def destroy_meal
+    @branch_meal.destroy ? flash[:notice] = t('.success') : flash[:alert] = t('.failure')
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/branches_controller.rb:67]: no need to destroy these, we can just deactivate the branch meals

File: app/controllers/branches_controller.rb [hunk 3]
```diff
@@ -34,11 +35,54 @@ def destroy
     redirect_to branches_url
   end
 
+  def meals
+    @branch_meals = BranchMeal.includes(:meal, :branch).where(branch_id: @branch.id).order(:id)
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/branches_controller.rb:39]: branch.meals

File: app/models/branch_meal.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+class BranchMeal < ApplicationRecord
+  self.table_name = "branches_meals"
+
+  ### ASSOCIATIONS ###
+
+  belongs_to :branch
+  belongs_to :meal
+  has_many :ingredients_meals, through: :meal
```
Comment (himanshumishra31) [inline app/models/branch_meal.rb:8]: dependent missing

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: app/models/branch_meal.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  has_many :inventories, through: :branch
+
+  ### VALIDATIONS ###
+
+  validates :meal_id, uniqueness: { scope: :branch_id }
+
+  ### CALLBACKS ###
+
+  after_save_commit :update_available
```
Comment (himanshumishra31) [inline app/models/branch_meal.rb:17]: add these callbacks in respective models.

File: app/models/ingredient_meal.rb [hunk 6]
```diff
@@ -5,6 +5,8 @@ class IngredientMeal < ApplicationRecord
 
   belongs_to :ingredient
   belongs_to :meal
+  has_many :branches_meals, through: :meal
```
Comment (himanshumishra31) [inline app/models/ingredient_meal.rb:8]: dependent missing

File: app/models/branch_meal.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+  def update_available
+    available_value = true
+    im = ingredients_meals.joins(:inventories).where(inventories: { branch_id: branch_id })
+    available_value = false if im.size < ingredients_meals.size || im.where('inventories.quantity < ingredients_meals.ingredient_quantity').any?
```
Comment (himanshumishra31) [inline app/models/branch_meal.rb:22]: check size vs count vs length

Comment (himanshumishra31) [reply app/models/branch_meal.rb:22]: add exists?

File: app/models/ingredient_meal.rb [hunk 8]
```diff
@@ -27,4 +31,21 @@ def price
   private def set_meal_non_veg
     meal.set_non_veg
   end
+
+  private def set_branch_meal_availability
+    if ingredient_quantity_previously_changed?
+      target_branches_meals = branches_meals.joins(:inventories).where('inventories.ingredient_id = ?', ingredient_id)
```
Comment (himanshumishra31) [inline app/models/ingredient_meal.rb:37]: call ingredient meal method

Comment (himanshumishra31) [reply app/models/ingredient_meal.rb:37]: if the request is taking too much time..We can add delayed jobs

**Review by himanshumishra31 — CHANGES_REQUESTED** [review]
make request changes

File: app/controllers/branches_controller.rb [hunk 9]
```diff
@@ -34,11 +35,54 @@ def destroy
...
+
+    if @branch_meal.save
+      redirect_to meals_branch_path(params[:id]), notice: t('.success')
+    else
+      render :add_meal, status: :unprocessable_entity
+    end
+  end
+
+  def toggle_meal_active
+    @branch_meal.update(active: true)
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/branches_controller.rb:57]: Done

File: app/models/branch_meal.rb [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  ### ASSOCIATIONS ###
+
+  belongs_to :branch
+  belongs_to :meal
+  has_many :ingredients_meals, through: :meal
+  has_many :inventories, through: :branch
+
+  ### VALIDATIONS ###
+
+  validates :meal_id, uniqueness: { scope: :branch_id }
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/branch_meal.rb:13]: Done

File: db/migrate/20230526110857_create_join_table_branch_meal.rb [hunk 11]
```diff
@@ -0,0 +1,10 @@
+class CreateJoinTableBranchMeal < ActiveRecord::Migration[7.0]
+  def change
+    create_table :branches_meals do |t|
+      t.references :branch
+      t.references :meal
+      t.boolean :active
+      t.boolean :available
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply db/migrate/20230526110857_create_join_table_branch_meal.rb:7]: Done

File: app/views/branches/_branch_meal.html.erb (hunk 1 above)
Comment (rishabh-vinsol) [reply app/views/branches/_branch_meal.html.erb:26]: SIr delete is for deleting branches_meals. I think it should be present.

Comment (rishabh-vinsol) [reply app/views/branches/_branch_meal.html.erb:26]: Done

File: app/controllers/branches_controller.rb (hunk 3 above)
Comment (rishabh-vinsol) [reply app/controllers/branches_controller.rb:39]: Done

File: app/models/ingredient_meal.rb (hunk 6 above)
Comment (rishabh-vinsol) [reply app/models/ingredient_meal.rb:8]: Discussed

File: app/models/branch_meal.rb (hunk 7 above)
Comment (rishabh-vinsol) [reply app/models/branch_meal.rb:22]: Done

File: app/models/ingredient_meal.rb (hunk 8 above)
Comment (rishabh-vinsol) [reply app/models/ingredient_meal.rb:37]: Did not make job for now. Updated

File: app/models/branch_meal.rb (hunk 5 above)
Comment (rishabh-vinsol) [reply app/models/branch_meal.rb:17]: Done

File: app/controllers/branches_controller.rb (hunk 2 above)
Comment (rishabh-vinsol) [reply app/controllers/branches_controller.rb:67]: Done

File: app/models/branch_meal.rb (hunk 4 above)
Comment (rishabh-vinsol) [reply app/models/branch_meal.rb:8]: Discussed

File: app/controllers/branches_controller.rb [hunk 12]
```diff
@@ -1,6 +1,7 @@
 # Controller to handle model Branch CRUD
 class BranchesController < ApplicationController
-  before_action :set_branch, only: %i[ show edit update destroy ]
+  before_action :set_branch, only: %i[ show edit update destroy meals add_meal create_meal ]
+  before_action :set_branch_meal, only: %i[ toggle_meal_active toggle_meal_inactive destroy_meal ]
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/branches_controller.rb:4]: We don't have this action now `destroy_meal `

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]
Getting this error when i am just editing inventory and clicking on save without changing the quantity
Minor, but pls fix
image

File: app/views/inventories/_form.html.erb [hunk 13]
```diff
@@ -2,7 +2,7 @@
   <%= render 'shared/errors', model: inventory, model_name: 'inventory' %>
   <div class="mb-3">
     <%= form.label :ingredient_id, 'Ingredient', class: "form-label" %>
-    <%= form.collection_select :ingredient_id, Ingredient.all, :id, :name, {prompt: "Select one"}, {class: "form-select"}%>
+    <%= form.collection_select :ingredient_id, Ingredient.all, :id, :name, {prompt: "Select one"}, {class: "form-select", disabled: ['edit', 'update'].include?(action_name) ? true : false }%>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/inventories/_form.html.erb:5]: Forgot to do that. include? returns true or false

File: app/views/branches/meals.html.erb [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+      <h1><%= link_to "Branches", branches_path, class: "text-decoration-none"%> / <%= @branch.name %> / Meals</h1>
+      <%= link_to "Add meal", add_meal_branch_path(@branch), class: "btn btn-primary btn-sm" %>
+    </div>
+    <% if @branch_meals.blank? %>
+      <div class="alert alert-dark mx-5" role="alert">
+        <strong>No meals present for this branch.</strong>
+      </div>
+    <% else %>
+      <div class="alert alert-light" role="alert">
+        This page is just to activate/deactivate meals is branch. To make changes in meals visit <%= link_to 'meals page', meals_path, class:'alert-link' %>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/branches/meals.html.erb:14]: Done

**Review by Sparsh-Goyal-2000 — APPROVED** [review]

**Pull request merged**